### ISSUE TITLE
Rate limiting does not respect dimensions without override

### ISSUE THREAD

**Issue by arnecls** [istio/istio#15958]
**Bug description**

When using rate limiting (redis) without defining an override, the key used to look up the current amount in redis does not respect any dimensions, unless there is an override provided.

See https://github.com/istio/istio/blob/89cb55316e79ecfa28b794298551299eef9da488/mixer/adapter/redisquota/redisquota.go#L292

Here the key is defined as the name of the quota used.
If there is an override, the name will be extended by a hash of the matching dimension.
If no override is provided, the key will be unchanged. See https://github.com/istio/istio/blob/89cb55316e79ecfa28b794298551299eef9da488/mixer/adapter/redisquota/redisquota.go#L300
Furthermore the key is the only identifier passed to the lua script, see https://github.com/istio/istio/blob/89cb55316e79ecfa28b794298551299eef9da488/mixer/adapter/redisquota/redisquota.go#L330

This is problematic, as instances referring to the same quota will effectively use the same counter _unless_ you have an override provided. I.e. different services might accidentally share the same counter.
Also this does not allow per-pod rate limiting. While you can provide a dimension containing `destination.name`, you cannot provide an override (as you cannot know all pod names), and thus all pods will use the same key.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure

**Expected behavior**

Even without providing an override, the dimensions defined in the instance (quota template) should generate unique keys.

**Steps to reproduce the bug**

1. Create a deployment and service called "test" in the namespace "default".
2. Scale this deployment to 1 replica
3. Apply the following config
4. Send about 10req/sec to the test service. It is limited to 1 req/sec
5. Increase the amount of replicas in test to 4 instances

You should see an increase to 4 req/sec (as dimensions use the pod-id), but it stays at 1 req/sec.

```yaml
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpecBinding
metadata:
  name: ratelimit
  namespace: search-core
spec:
  quotaSpecs:
  - name: ratelimit
    namespace: istio-system
  services:
  - name: test
    namespace: default
---
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: redis-quota-handler
  namespace: istio-system
spec:
  compiledAdapter: redisquota
  params:
    redisServerUrl: 10.246.16.131:6379
    connectionPoolSize: 10
    quotas:
    - name: ratelimit.instance.istio-system
      maxAmount: 1
      validDuration: 1s
      bucketDuration: 500ms
      rateLimitAlgorithm: ROLLING_WINDOW
---
apiVersion: config.istio.io/v1alpha2
kind: instance
metadata:
  name: ratelimit
  namespace: istio-system
spec:
  compiledTemplate: quota
  params:
    dimensions:
      destination: destination.name
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpec
metadata:
  name: ratelimit
  namespace: istio-system
spec:
  rules:
  - quotas:
    - charge: 1
      quota: ratelimit
---
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: ratelimit-grpc
  namespace: istio-system
spec:
  actions:
  - handler: redis-quota-handler
    instances:
    - ratelimit
```

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

Istio 1.2.2 (installed via helm) on GKE 1.13

**Comment by arnecls** [comment]
One thing that would (probably) solve it, but is very slow:

```go
func (h *handler) getKeyAndQuotaAmount(instance *quota.Instance, quota *config.Params_Quota) (string, int64, error) {
	maxAmount := quota.MaxAmount
	key := quota.Name

// for loop to cover override ...

	id := make(map[string]string)
	for k, v := range instance.Dimensions {
		switch s := v.(type) {
		case string:
			id[k] = s
		case fmt.Stringer:
			id[k] = s.String()
	}
	key = key + "-" + getDimensionHash(id)

	return key, maxAmount, nil
}
```

It's a bit sad, that instance.dimensions is a `map[string]interface{}` so we have to do this string conversion here (not sure if this will work at all for non-string types, as I don't know if they are all stringers). Also I don't see a way of caching the hash for a specific instance to speed up this process.

**Comment by douglas-reid** [comment]
There were several PRs related to fixing the redisquota dimensioning.  Have you verified this with latest `master`?

**Comment by douglas-reid** [comment]
i believe this has been fixed, and should be in the next 1.2.x release and the 1.3 release. Closing.

please re-open if this is not the case.

**Comment by arnecls** [comment]
I checked the code and on a first glance it looks ok. 👍 
I will do a proper test once there is a new release with this  change included.
